secant/Util/RecoveryPhraseStorage.swift

```diff
@@ -0,0 +1,228 @@
+            throw KeychainError.noDataFound
+        }
+        
+        return String(decoding: data, as: UTF8.self)
+    }
+
+    func setRecoveryPhrase(_ phrase: String) throws {
+        guard let data = phrase.data(using: .utf8) else {
+            throw KeychainError.encoding
+        }
+            
+        try setData(data, forKey: Constants.zcashRecoveryPhrase)
+    }
+    
+    func birthday() throws -> BlockHeight {
+        guard let data = data(forKey: Constants.zcashBirthday) else {
+            throw KeychainError.noDataFound
+        }
+
+        do {
+            if let birthday = try NSKeyedUnarchiver.unarchivedObject(ofClass: NSNumber.self, from: data) {
+                return birthday.intValue
+            }
+            
+            throw KeychainError.decoding
+        } catch {
+            throw error
+        }
+    }
+
+    /// Use carefully:  Deletes birthday from the keychain!!!
+    @discardableResult
+    func deleteBirthday() -> Bool {
+        deleteData(forKey: Constants.zcashBirthday)
+    }
+
+    func setBirthday(_ birthday: BlockHeight) throws {
+        do {
+            let data = try NSKeyedArchiver.archivedData(withRootObject: NSNumber(value: birthday), requiringSecureCoding: true)
+            
+            try setData(data, forKey: Constants.zcashBirthday)
+        } catch {
+            throw error
+        }
+    }
+    
+    /// Restore data for key
+    func data(
+        forKey: String,
+        account: String = ""
+    ) -> Data? {
+        let query = RecoveryPhraseStorage.restoreQuery(forAccount: account, andKey: forKey)
+
+        var result: AnyObject?
+        _ = SecItemCopyMatching(query as CFDictionary, &result)
+        
+        return result as? Data
+    }
+    
+    /// Use carefully:  Deletes data for key
+    func deleteData(
+        forKey: String,
+        account: String = ""
+    ) -> Bool {
+        let query = RecoveryPhraseStorage.baseQuery(forAccount: account, andKey: forKey)
+
+        let status = SecItemDelete(query as CFDictionary)
+
+        return status == noErr
+    }
+    
+    /// Store data for key
+    func setData(
+        _ data: Data,
+        forKey: String,
+        account: String = ""
+    ) throws {
+        var query = RecoveryPhraseStorage.baseQuery(forAccount: account, andKey: forKey)
+        query[kSecValueData as String] = data as AnyObject
+
+        let status = SecItemAdd(query as CFDictionary, nil)
+        
+        guard status != errSecDuplicateItem else {
+            throw KeychainError.duplicate
+        }
+        
+        guard status == errSecSuccess else {
+            throw KeychainError.unknown(status)
+        }
+    }
+}
+
+// MARK: - KeyStoring
+
+extension RecoveryPhraseStorage: KeyStoring {
+    func importBirthday(_ height: BlockHeight) throws {
+        do {
+            try setBirthday(height)
+        } catch RecoveryPhraseStorage.KeychainError.duplicate {
+            throw KeyStoringError.alreadyImported
```

Comment:
Will the default birthday be stored during wallet import if there is none provided by the user?  I want to clarify what our representation invariants are, and how we deal with updating the birthday later.

A possible scenario:
 - User does a wallet import
 - Birthday is optional, so user doesn't put one in.
 - App could either write nothing for birthday, or it could write the network default.  This is currently unspecified behavior
 - Later when the blockchain is scanned, it can update the birthday with the known correct value.  The main use would be for re-display to the user later.  This would necessitate writing the birthday which implies the prior step would need to write only the seed and not the birthday.

And what happens if the user puts in a bad birthday?  When we scan the blockchain, will we need to update the birthday with a corrected value?

Reply:
This is a really good observation. There are many things to this point. When restoring from. recovery phrase, birthday is optional, since we don't know what's the origin of it. BIP-39 will give us reassurance of its integrity and ability to reconstruct the seed bytes, it won't prescript anything regarding wallet birthday. 

The user might not have the birthday or be unsure about it. So If it doesn't have it the wallet should assume sapling activation for syncing height. 

The ECC reference wallets differed in this point. The iOS wallet stored the resulting birthday from the restore process, that being the one that the uses wrote in, or sapling activation. The Android wallet does something additional which is updating the birthday to the least amount of blocks as possible. 

What does this mean?

Providing a birthday doesn't mean that the wallet has been initiated at that block height. The first assumption is that the wallet won't have any transactions involving those keys prior to the height signaled as birthday. The second assumption is that the wallet factually starts at that given height.

The first assumption is rare and expensive to verify, the second one could be conveniently verified when the wallet syncs past that given height.

The Android app does this. Even thought it accepts the birthday that the user provides as input, it overwrites it with a more recent block height if it has been found that the first received transaction has a greater height than the one stored in the secure storage. 

Although this is convenient, it can be confusing if it's not communicated properly to the user. I, as a. user, can be shown a backup screen that has a different birthday than the one I remember having backed up.

Reply:
is there any outcome of this discussion please?

Reply:
Let's got with storing `nil` instead of default `saplingActivation` for when no birthday is provided